### PR TITLE
feat: snapbox migration

### PR DESCRIPTION
**Pull request by zerosnacks**
## Motivation

The goal of the PR is to sufficiently address the request outlined in: 

Closes: https://github.com/foundry-rs/foundry/issues/8389

Migrates most tests to use Snapbox.

Cleans up a lot of old helpers. Ideally we can get rid of `stdout_lossy()` all together at some point.

Disallows `not_empty` pattern, favoring Snapbox. Helper internals have been refactored to use Snapbox.

## Solution

Improves a number of tests by applying more thorough match assertions

In order to port the last test cases / make testing more ergonomic:

**Known issues**:

- `stdout_lossy` is still used in contexts of capturing the value from two commands and determine equality or casting to a value is required to make a comparison:

https://github.com/foundry-rs/foundry/blob/bdf48aa6f04f69f40c8eb3623cf32f7413a82ea4/crates/cast/tests/cli/main.rs#L1021-L1025

- There is currently no good way to capture the output (to do the above) and perform the regular `stdout_eq` check
- Sometimes ordering of output cannot be enforced, it is only possible to use `.contains` on the output here:

https://github.com/foundry-rs/foundry/blob/bdf48aa6f04f69f40c8eb3623cf32f7413a82ea4/crates/cast/tests/cli/main.rs#L331-L393

**Suggestions on what we should add / request upstream**:

- `file!` should support dynamically constructed Path's or default to `CARGO_MANIFEST_DIR`, e.g. `PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../testdata/etherscan")`
- There is currently no way to dynamically construct the data passed to `stdout_eq`, it has to be statically defined. It is therefore not possible to embed constants inside of the argument passed to `stdout_eq`. It is also not possible to have an array of static data passed to `stdout_eq` in a loop - it must be unrolled manually.

### CODE REVIEW

File: crates/forge/tests/cli/build.rs [hunk 1]
```diff
@@ -55,10 +55,14 @@ Compiler run successful!
...
-    let stdout = cmd.stdout_lossy();
-    assert!(!stdout.contains("console"), "\n{stdout}");
-    assert!(!stdout.contains("std"), "\n{stdout}");
-    assert!(stdout.contains("Counter"), "\n{stdout}");
+    cmd.assert_success().stdout_eq(str![
+        r#"
+...
+| Contract | Size (B) | Margin (B) |
+|----------|----------|------------|
+...
```
Comment (DaniPopes) [inline crates/forge/tests/cli/build.rs:63]: Should include Counter

File: crates/test-utils/src/util.rs [hunk 2]
```diff
@@ -1100,6 +1092,8 @@ fn test_redactions() -> snapbox::Redactions {
             ("[FILE]", r"-->.*\.sol"),
             ("[FILE]", r"Location(.|\n)*\.rs(.|\n)*Backtrace"),
             ("[TX_HASH]", r"Transaction hash: 0x[0-9A-Fa-f]{64}"),
+            ("[ADDRESS]", r"Address: 0x[0-9A-Fa-f]{40}"),
+            ("[LIB_DIR]", r"/tmp/[^/]+-[^/]+/lib"),
```
Comment (DaniPopes) [inline crates/test-utils/src/util.rs:1096]: This is platform specific, should grep for "updating ..."

File: crates/forge/tests/cli/build.rs (hunk 1 above)
Comment (zerosnacks) [reply crates/forge/tests/cli/build.rs:64]: Done, due to formatting can't easily add the `[..]` into the table due to whitespace formatting but I can add the `Counter` title.

File: crates/cast/tests/cli/main.rs [hunk 3]
```diff
@@ -522,35 +621,45 @@ casttest!(receipt_revert_reason, |_prj, cmd| {
...
+    cmd.args([
         "receipt",
         "0x44f2aaa351460c074f2cb1e5a9e28cbc7d83f33e425101d2de14331c7b7ec31e",
         "--rpc-url",
         rpc.as_str(),
-    ]);
+    ])
+    .assert_success();
     let output = cmd.stdout_lossy();
     assert!(!output.contains("revertReason"));
```
Comment (zerosnacks) [inline crates/cast/tests/cli/main.rs:632]: It is currently not possible to make a negative assertion like this with Snapbox I think

File: crates/cast/tests/cli/main.rs [hunk 4]
```diff
@@ -584,11 +707,24 @@ casttest!(logs_topics, |_prj, cmd| {
...
+  	0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef
+  	0x000000000000000000000000ab5801a7d398351b8be11c439e05c5b3259aec9b
+  	0x00000000000000000000000068a99f89e475a078645f4bac491360afe255dff1
+  ]
+  transactionHash: 0xb65bcbb85c1633b0ab4e4886c3cd8eeaeb63edbb39cacdb9223fdcf4454fd2c7
+  transactionIndex: 8
 
-    cmd.unchecked_output().stdout_matches_path(
-        Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/fixtures/cast_logs.stdout"),
-    );
```
Comment (zerosnacks) [inline crates/cast/tests/cli/main.rs:591]: Notice `file!` cannot be used here as currently implemented as it only accepts statically defined strings

File: crates/cast/tests/cli/main.rs [hunk 5]
```diff
@@ -584,11 +707,24 @@ casttest!(logs_topics, |_prj, cmd| {
...
+  data: 0x000000000000000000000000000000000000027fd7b375dda5ef932dac18d302
+  logIndex: 15
+  removed: false
+  topics: [
+  	0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef
+  	0x000000000000000000000000ab5801a7d398351b8be11c439e05c5b3259aec9b
+  	0x00000000000000000000000068a99f89e475a078645f4bac491360afe255dff1
+  ]
+  transactionHash: 0xb65bcbb85c1633b0ab4e4886c3cd8eeaeb63edbb39cacdb9223fdcf4454fd2c7
+  transactionIndex: 8
```
Comment (zerosnacks) [inline crates/cast/tests/cli/main.rs:725]: notice this should be ideally we moved to a constant (or file) as it is the same for the following cases

unfortunately this does not appear to be possible, only `inline` is supported - leading to duplicate inlined cases

File: crates/cast/tests/cli/main.rs [hunk 6]
```diff
@@ -1054,14 +1051,16 @@ casttest!(storage_layout, |_prj, cmd| {
 casttest!(balance, |_prj, cmd| {
```
Comment (DaniPopes) [inline crates/cast/tests/cli/main.rs:1051]: this test is malformed, .args() doesn't execute the command so the two variables are the same

File: crates/cast/tests/cli/main.rs (hunk 3 above)
Comment (DaniPopes) [reply crates/cast/tests/cli/main.rs:632]: the point is that the output is shown in full and any change fails the test

File: crates/cast/tests/cli/main.rs (hunk 5 above)
Comment (DaniPopes) [reply crates/cast/tests/cli/main.rs:725]: it's `impl IntoData`, you can use file![] for all the cases as before I think?

Comment (zerosnacks) [reply crates/cast/tests/cli/main.rs]: `.stdout_eq(file!["../fixtures/cast_logs.stdout"])` is supported whereas `.stdout_eq(file![Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/fixtures/cast_logs.stdout")]);` is not

Updated the cases to use the `file!` macro but with relative paths, this does not appear to cause issues

**Review by mattsse — APPROVED** [review]
lgtm, pending @DaniPopes

File: crates/forge/tests/cli/build.rs [hunk 7]
```diff
@@ -45,7 +45,7 @@ contract Dummy {
 // tests build output is as expected
 forgetest_init!(exact_build_output, |prj, cmd| {
     cmd.args(["build", "--force"]).assert_success().stdout_eq(str![[r#"
-Compiling 27 files with [SOLC_VERSION]
+Compiling [..] files with [SOLC_VERSION]
```
Comment (DaniPopes) [inline crates/forge/tests/cli/build.rs:48]: Maybe redact this by default too?

**Review by DaniPopes — APPROVED** [review]
great

File: crates/forge/tests/cli/build.rs (hunk 7 above)
Comment (zerosnacks) [reply crates/forge/tests/cli/build.rs]: Done

**Pull request merged**